cpp/src/join/filtered_join.cu

```diff
@@ -0,0 +1,405 @@
+
+#include <algorithm>
+#include <memory>
+#include <type_traits>
+
+namespace cudf {
+namespace detail {
+namespace {
+
+/**
+ * @brief Build a row bitmask for the input table.
+ *
+ * The output bitmask will have invalid bits corresponding to the input rows having nulls (at
+ * any nested level) and vice versa.
+ *
+ * @param input The input table
+ * @param stream CUDA stream used for device memory operations and kernel launches
+ * @return A pair of pointer to the output bitmask and the buffer containing the bitmask
+ */
+std::pair<rmm::device_buffer, bitmask_type const*> build_row_bitmask(table_view const& input,
+                                                                     rmm::cuda_stream_view stream)
+{
+  auto const nullable_columns = get_nullable_columns(input);
+  CUDF_EXPECTS(nullable_columns.size() > 0,
+               "The input table has nulls thus it should have nullable columns.");
+
+  // If there are more than one nullable column, we compute `bitmask_and` of their null masks.
+  // Otherwise, we have only one nullable column and can use its null mask directly.
+  if (nullable_columns.size() > 1) {
+    auto row_bitmask =
+      cudf::detail::bitmask_and(
+        table_view{nullable_columns}, stream, cudf::get_current_device_resource_ref())
+        .first;
+    auto const row_bitmask_ptr = static_cast<bitmask_type const*>(row_bitmask.data());
+    return std::pair(std::move(row_bitmask), row_bitmask_ptr);
+  }
+
+  return std::pair(rmm::device_buffer{0, stream}, nullable_columns.front().null_mask());
+}
+
+}  // namespace
+
+auto filtered_join::compute_bucket_storage_size(cudf::table_view tbl, double load_factor)
+{
+  return std::max(
+    {static_cast<cudf::size_type>(
+       cuco::make_valid_extent<primitive_probing_scheme, storage_type, cudf::size_type>(
+         tbl.num_rows(), load_factor)),
+     static_cast<cudf::size_type>(
+       cuco::make_valid_extent<nested_probing_scheme, storage_type, cudf::size_type>(tbl.num_rows(),
+                                                                                     load_factor)),
+     static_cast<cudf::size_type>(
+       cuco::make_valid_extent<simple_probing_scheme, storage_type, cudf::size_type>(
+         tbl.num_rows(), load_factor))});
+}
+
+template <int32_t CGSize, typename Ref>
+void filtered_join::insert_build_table(Ref const& insert_ref, rmm::cuda_stream_view stream)
+{
+  nvtxRangePushA("insert_build_table");
+  auto insert = [&]<typename Iterator>(Iterator build_iter) {
+    // Build hash table by inserting all rows from build table
+    auto const grid_size = cuco::detail::grid_size(_build.num_rows(), CGSize);
+
+    if (_build_props._has_nulls && _nulls_equal == null_equality::UNEQUAL) {
+      auto const bitmask_buffer_and_ptr = build_row_bitmask(_build, stream);
+      auto const row_bitmask_ptr        = bitmask_buffer_and_ptr.second;
+      cuco::detail::open_addressing_ns::insert_if_n<CGSize, cuco::detail::default_block_size()>
```

Comment:
Can we pass in the has table instead, so we can just call `hash_table.insert_if_async`?

Reply:
This idea comes directly from this PR, allowing `filtered_join` to avoid a template parameter. As you may notice, instead of creating a cuco hash table, we initialize only the hash table storage during construction and create hash table refs on the fly for insertions and queries. This enables using different CG sizes for nested and non-nested types without complicating the public interface.